iree/hal/cuda/registration/driver_module.cc

```diff
@@ -0,0 +1,69 @@
+// Copyright 2021 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "iree/hal/cuda/registration/driver_module.h"
+
+#include <inttypes.h>
+
+#include "iree/base/status.h"
+#include "iree/base/target_platform.h"
+#include "iree/base/tracing.h"
+#include "iree/hal/cuda/api.h"
+
+#define IREE_HAL_CUDA_DRIVER_ID 0x43554441u  // CUDA
+
+static iree_status_t iree_hal_cuda_driver_factory_enumerate(
+    void* self, const iree_hal_driver_info_t** out_driver_infos,
+    iree_host_size_t* out_driver_info_count) {
+  // NOTE: we could query supported cuda versions or featuresets here.
+  static const iree_hal_driver_info_t driver_infos[1] = {{
+      /*driver_id=*/IREE_HAL_CUDA_DRIVER_ID,
+      /*driver_name=*/iree_make_cstring_view("cuda"),
+      /*full_name=*/iree_make_cstring_view("CUDA (dynamic)"),
+  }};
+  *out_driver_info_count = IREE_ARRAYSIZE(driver_infos);
+  *out_driver_infos = driver_infos;
+  return iree_ok_status();
+}
+
+static iree_status_t iree_hal_cuda_driver_factory_try_create(
+    void* self, iree_hal_driver_id_t driver_id, iree_allocator_t allocator,
+    iree_hal_driver_t** out_driver) {
+  if (driver_id != IREE_HAL_CUDA_DRIVER_ID) {
```

Comment:
(continuing from below, about default initializing)
Add `*out_driver = NULL;` here - without that, *out_driver is uninitialized upon this failure here. Callers *should* be checking the status and not using it, but that's a cheap assumption to avoid and not have to worry about. Here and elsewhere. I've tried to keep a convention of asserting on non-null params and initializing and it'd be nice to carry that over here:

```c
IREE_ASSERT_ARGUMENT(self);
IREE_ASSERT_ARGUMENT(out_driver);
*out_driver = NULL;
<parameter validation that may return failures>
IREE_TRACE_ZONE_BEGIN(z0);
iree_status_t status = <actual meat of the function>;
if (iree_status_is_ok(status)) {
  *out_driver = driver;
} else {
  iree_hal_driver_release(driver);
}
IREE_TRACE_ZONE_END(z0);
return status;
```

then you get nice easy to spot assertions for null arguments vs requiring ASAN (which is not available everywhere) or getting random heap corruptions as you dereference NULL+0x12.  You also avoid the need for a bunch of trace handling at the start as argument validation - unless it's doing something crazy - is not usually meaningful to trace, and only if the actual work the function is doing is allocation you don't even need a lot of the other stuff too. Consider the above as the boilerplate all functions start with and then you remove what's not needed. So if you had no out arguments, you'd have:

```c
IREE_ASSERT_ARGUMENT(self);
<parameter validation that may return failures>
IREE_TRACE_ZONE_BEGIN(z0);
iree_status_t status = <actual meat of the function>;
IREE_TRACE_ZONE_END(z0);
return status;
```

or no tracing (it's an accessor or something simple):
```c
IREE_ASSERT_ARGUMENT(self);
<parameter validation that may return failures>
return <actual meat of the function>;
```

etc etc

(I'm really just priming things for a conventions guide here - it's super helpful to work through this with you :)